File: C++/data-structures/graphs.cpp
#include <iostream>
#include <string>
#include <vector>

class Node
{
public:
    std::string name;
    std::vector<Node *> children;

    Node(std::string str) { name = str; }

    std::vector<std::string> depthFirstSearch(std::vector<std::string> *array)
    {
        array->push_back(this->name);
        for (int i = 0; i < this->children.size(); i++)
        {
            children[i]->depthFirstSearch(array);
        }
        return *array;
    }

    Node *addChild(std::string name)
    {
        Node *child = new Node(name);
        children.push_back(child);
        return this;
    }
};

int main()
{
    Node graph("A");
    graph.addChild("B")->addChild("C")->addChild("D");
    graph.children[0]->addChild("E")->addChild("F");
    graph.children[2]->addChild("G")->addChild("H");
    graph.children[0]->children[1]->addChild("I")->addChild("J");
    graph.children[2]->children[0]->addChild("K");

    std::vector<std::string> result;
    graph.depthFirstSearch(&result);

    for (int i = 0; i < result.size(); i++)
    {
        std::cout << result[i] << " ";
    }
    std::cout << std::endl;

    return 0;
}

File: Go/data-structures/graphs.go
package data_structures

type Node struct {
	Name     string
	Children []*Node
}

func NewNode(name string) *Node {
	return &Node{
		Name:     name,
		Children: []*Node{},
	}
}

func (n *Node) AddChildren(names ...string) *Node {
	for _, name := range names {
		child := Node{Name: name}
		n.Children = append(n.Children, &child)
	}
	return n
}

// Traverse
// Depth First Search (DFS)
func (n *Node) DepthFirstSearch(array *[]string, visited map[string]bool) []string {
	*array = append(*array, n.Name)
	visited[n.Name] = true
	for _, child := range n.Children {
		if _, ok := visited[child.Name]; !ok {
			visited[child.Name] = true
			*array = child.DepthFirstSearch(array, visited)
		}
	}
	return *array
}

// DFS for acyclic graphs or Trees
//func (n *Node) DepthFirstSearch(array []string) []string {
//	array = append(array, n.Name)
//	for _, child := range n.Children {
//			array = child.DepthFirstSearch(array)
//	}
//	return array
//}
